src/repr/src/row.rs

```diff
@@ -1876,16 +1888,73 @@ impl RowPacker<'_> {
     where
         F: FnOnce(&mut RowPacker) -> R,
     {
-        self.row.data.push(Tag::List.into());
+        // First, assume that the list will fit in 255 bytes, and thus the length will fit in
+        // 1 byte. If not, we'll fix it up later.
         let start = self.row.data.len();
-        // write a dummy len, will fix it up later
-        self.row.data.extend_from_slice(&[0; size_of::<u64>()]);
+        self.row.data.push(Tag::ListTiny.into());
+        // Write a dummy len, will fix it up later.
+        self.row.data.push(0);
 
         let out = f(self);
 
-        let len = u64::cast_from(self.row.data.len() - start - size_of::<u64>());
-        // fix up the len
-        self.row.data[start..start + size_of::<u64>()].copy_from_slice(&len.to_le_bytes());
+        // The `- 1 - 1` is for the tag and the len.
+        let len = self.row.data.len() - start - 1 - 1;
+        // We now know the real len.
+        if len <= 255 {
+            // If the len fits in 1 byte, we just need to fix up the len.
+            self.row.data[start + 1] = len.to_le_bytes()[0];
+        } else {
+            // Note: We move this code path into its own function, so that the common case can be
+            // inlined.
+            long_list(&mut self.row.data, start, len);
+        }
+
+        /// 1. Fix up the tag.
+        /// 2. Move the actual data a bit (for which we also need to make room at the end).
+        /// 3. Fix up the len.
+        /// `data`: The row's backing data.
+        /// `start`: where `push_list_with` started writing in `data`.
+        /// `len`: the length of the data, excluding the tag and the length.
+        #[cold]
+        fn long_list(data: &mut CompactBytes, start: usize, len: usize) {
+            match len {
+                0..=255 => {
+                    unreachable!()
+                }
+                256..=65535 => {
+                    data[start] = Tag::ListShort.into();
+                    long_list_inner(data, start, len, 2);
+                }
+                65536..=4294967295 => {
+                    data[start] = Tag::ListLong.into();
+                    long_list_inner(data, start, len, 4);
+                }
+                _ => {
+                    data[start] = Tag::ListHuge.into();
+                    long_list_inner(data, start, len, 8);
+                }
+            };
+            /// `len_len`: the length of the length. (Possible values are: 2, 4, 8. 1 is handled
+            /// elsewhere.)
+            /// The other parameters are the same as for `long_list`.
+            fn long_list_inner(data: &mut CompactBytes, start: usize, len: usize, len_len: usize) {
```

Comment:
By "parameters get inlined", do you mean that with a closure, we can capture `start` and `len` instead of being passed in as a parameter? (Note that the closure can't capture `data`, because then the closure would need to borrow the `&mut data`, and I'm also modifying `data` outside the closure.)

I've changed to a closure that captures `start` and `len`.